test/src/test/java/jenkins/security/ResourceDomainTest.java

```diff
@@ -0,0 +1,149 @@
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.For;
+import org.jvnet.hudson.test.Issue;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.MockAuthorizationStrategy;
+
+import java.net.URL;
+import java.util.UUID;
+
+@Issue("JENKINS-41891")
+@For({ ResourceDomainRootAction.class, ResourceDomainFilter.class, ResourceDomainConfiguration.class })
+public class ResourceDomainTest {
+
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    private static final String RESOURCE_DOMAIN = "127.0.0.1";
+
+    @Before
+    public void prepare() throws Exception {
+        String resourceRoot;
+        URL root = j.getURL(); // which always will use "localhost", see JenkinsRule#getURL()
+
+        resourceRoot = root.toString().replace("localhost", RESOURCE_DOMAIN);
+        ResourceDomainConfiguration configuration = ExtensionList.lookupSingleton(ResourceDomainConfiguration.class);
+        configuration.setUrl(resourceRoot);
+    }
+
+    @Test
+    public void secondDomainBasics() throws Exception {
+        JenkinsRule.WebClient webClient = j.createWebClient();
+
+        { // DBS on primary domain forwards to second domain when trying to access a file URL
+            Page page = webClient.goTo("userContent");
+            Assert.assertEquals("successful request", 200, page.getWebResponse().getStatusCode());
+            Assert.assertTrue("web page", page.isHtmlPage());
+        }
+
+        String resourceResponseUrl;
+        {
+            webClient.setRedirectEnabled(true);
+            Page page = webClient.goTo("userContent/readme.txt", "text/plain");
+            resourceResponseUrl = page.getUrl().toString();
+            Assert.assertEquals("resource response success", 200, page.getWebResponse().getStatusCode());
+            Assert.assertNull("no CSP headers", page.getWebResponse().getResponseHeaderValue("Content-Security-Policy"));
+            Assert.assertTrue("Served from resource domain", resourceResponseUrl.contains(RESOURCE_DOMAIN));
+            Assert.assertTrue("Served from resource action", resourceResponseUrl.contains("static-files"));
+        }
+
+        { // direct access to resource URL works
+            Page page = webClient.getPage(resourceResponseUrl);
+            resourceResponseUrl = page.getUrl().toString();
+            Assert.assertEquals("resource response success", 200, page.getWebResponse().getStatusCode());
+            Assert.assertNull("no CSP headers", page.getWebResponse().getResponseHeaderValue("Content-Security-Policy"));
+            Assert.assertTrue("Served from resource domain", resourceResponseUrl.contains(RESOURCE_DOMAIN));
+            Assert.assertTrue("Served from resource action", resourceResponseUrl.contains("static-files"));
+        }
+
+        { // TODO This test might actually be an undesirable outcome for DBS that set an index file URL, TBD
```

Comment:
Javadoc browsing as of https://github.com/jenkinsci/javadoc-plugin/pull/4 is broken; perhaps this is why?

Reply:
FTR

```groovy
podTemplate(containers: [containerTemplate(name: 'maven', image: 'maven', command: 'sleep', args: '99d')]) {
  node(POD_LABEL) {
    git 'https://github.com/apache/commons-io'
    container('maven') {
      sh 'mvn -B javadoc:javadoc-no-fork'
      step([$class: 'JavadocArchiver', javadocDir: 'target/site/apidocs'])
    }
  }
}
```

Reply:
Works for me after recent updates; it was the behavior called out in this TODO as I expected.